File: track_chopstick.py
import cv2 as cv
import numpy as np
import time

def make_mask_image(img_bgr):
    img_hsv = cv.cvtColor(img_bgr, cv.COLOR_BGR2HSV)

    # img_h,img_s,img_v = cv.split(img_hsv)

    low = (0, 20, 0)

    high = (25, 255, 255)

    img_mask = cv.inRange(img_hsv, low, high)
    img_mask = cv.bitwise_not(img_mask)
    return img_mask

def detect_chopstic(img, thr, minLineLength, maxLineGap):
    cur_line=[]
    edges = cv.Canny(img, 50, 150, apertureSize=3)

    # lines = cv.HoughLines(edges, 1,np.pi/180,thr)
    lines = cv.HoughLinesP(edges, 1, np.pi / 180, thr, minLineLength=minLineLength, maxLineGap=maxLineGap)

    if lines is None:
        return None, None, None, None

    x1, y1, x2, y2 = lines[0][0]
    return x1, y1, x2, y2


def detect_spoon(img):
    circles = cv.HoughCircles(img, cv.HOUGH_GRADIENT, 1, 10,
                              param1=50, param2=12, minRadius=20, maxRadius=60)
    if circles is None:
        return None, None, None
    x = np.uint16(np.around(circles))[0][0][0]
    y = np.uint16(np.around(circles))[0][0][1]
    r = np.uint16(np.around(circles))[0][0][2]
    return x, y, r


cap = cv.VideoCapture(-1)
ret = cap.set(3, 640)
ret = cap.set(4, 480)
fgbg = cv.createBackgroundSubtractorMOG2(varThreshold=300, detectShadows=0)

prevTime = 0
while (1):
    ret, frame = cap.read()
    if ret == False:
        break
    curTime = time.time()
    sec = curTime - prevTime
    prevTime = curTime
    fps = 1 / (sec)
    print("Time {0} ".format(sec))
    print("Estimated fps {0} ".format(fps))
    str = "FPS : %0.1f" % fps
    frame = cv.flip(frame, 1)
    # 살색 제거 마스크 이미지
    mask_img = make_mask_image(frame)
    cv.imshow("test",mask_img)
    # 차영상 검출
    blur = cv.GaussianBlur(frame, (5, 5), 0)
    fgmask = fgbg.apply(blur, learningRate=0)
    kernel = cv.getStructuringElement(cv.MORPH_ELLIPSE, (5, 5))
    fgmask = cv.morphologyEx(fgmask, cv.MORPH_CLOSE, kernel, 2)

    x1,y1,x2,y2 = detect_chopstic(fgmask, 90, 15, 100)
    fgmask = cv.bitwise_and(mask_img, fgmask)
    fgmask = cv.medianBlur(fgmask,9)
    cv.imshow('mask', mask_img)

    # 젓가락 끝점 좌표
    x1, y1, x2, y2 = detect_chopstic(fgmask, 90, 15, 100)
    #c_x, c_y, r = detect_spoon(fgmask)
    # 직선이 검출되면 선분과 끝점 그리기
    if x1 != None:
        cv.line(frame, (x1, y1), (x2, y2), (0, 255, 0), 5)
        cv.circle(frame, (x1, y1), 10, (255, 0, 0), -1)
    # 원이 검출되면 원 그리기
    # if c_x != None:
    #     cv.circle(frame, (c_x, c_y), r, (0, 255, 0), 5)
    #     cv.circle(frame, (c_x, c_y), 10, (255, 0, 0), -1)
    # 원본 이미지
    cv.putText(frame, str, (0, 100), cv.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0))
    cv.imshow('img', frame)
    # 차영상
    cv.imshow('result', fgmask)

    key = cv.waitKey(30) & 0xff
    if key == 27:
        break

cap.release()
cv.destroyAllWindows()

File: track_chopstick.cpp

#include <opencv2/opencv.hpp>
#include <iostream>
#include <stdio.h>

using namespace cv;
using namespace std;



Vec4i detect_chopstic(Mat img, int thr, int minLineLength, int maxLineGap)
{
	Mat img_canny;
	Canny(img, img_canny, 50, 150);

	vector<Vec4i> lines;
	HoughLinesP(img_canny, lines, 1, CV_PI / 180, thr, minLineLength, maxLineGap);

	if (lines.empty())
	{
		return NULL;
	}

	return lines[0];
}

Vec3f detect_spoon(Mat img, int minDistance, int thr)
{
	Mat img_canny;
	Canny(img, img_canny, 50, 150);
	imshow("canny", img_canny);
	vector<Vec3f> circles;
	HoughCircles(img_canny, circles, HOUGH_GRADIENT, 1, minDistance, thr, thr / 2, 10, 50);

	if (circles.empty())
	{
		return NULL;
	}

	return circles[0];
}

Mat make_mask_image(Mat frame)
{
	Mat img_YCrCb;
	Mat mask;
	cvtColor(frame, img_YCrCb, COLOR_BGR2YCrCb);
	inRange(img_YCrCb, Scalar(0, 133, 77), Scalar(255, 173, 127), mask);
	//vector<Mat> planes;
	//split(img_YCrCb, planes);
	//Mat mask = (128 < planes[1]) & (planes[1] < 170) & (73 < planes[2]) & (planes[2] < 158);

	bitwise_not(mask, mask);
	return mask;
}

int main(int, char**)
{
	Mat frame, blur, mask_img, foregroundMask;
	//--- INITIALIZE VIDEOCAPTURE
	VideoCapture cap;
	// open the default camera using default API
	cap.open(0);
	// OR advance usage: select any API backend
	int deviceID = 0;             // 0 = open default camera
	int apiID = cv::CAP_ANY;      // 0 = autodetect default API
								  // open selected camera using selected API

	cap.open(deviceID + apiID);
	cap.set(CAP_PROP_AUTOFOCUS, 0);
	// check if we succeeded
	if (!cap.isOpened()) {
		cerr << "ERROR! Unable to open camera\n";
		return -1;
	}
	Ptr<BackgroundSubtractor> bg_model = createBackgroundSubtractorMOG2(0, 300, false);

	//--- GRAB AND WRITE LOOP
	cout << "Start grabbing" << endl;
	for (;;)
	{
		// wait for a new frame from camera and store it into 'frame'
		cap.read(frame);
		resize(frame, frame, Size(640, 480));
		flip(frame, frame, 1);
		GaussianBlur(frame, blur, Size(5, 5), 0);
		// check if we succeeded
		if (frame.empty()) {
			cerr << "ERROR! blank frame grabbed\n";
			break;
		}
		// create foreground mask of proper size
		if (foregroundMask.empty()) {
			foregroundMask.create(frame.size(), frame.type());
		}


		mask_img = make_mask_image(frame);
		imshow("mask_img", mask_img);
		bg_model->apply(blur, foregroundMask, false);
		///////////////////////////////////////////////////////////////
		Mat kernel = getStructuringElement(MORPH_RECT, Size(5, 5), Point(-1, -1));
		morphologyEx(foregroundMask, foregroundMask, MORPH_CLOSE, kernel);


		bitwise_and(mask_img, foregroundMask, foregroundMask);
		medianBlur(foregroundMask, foregroundMask, 9);
		imshow("mask", foregroundMask);

		//detect chopstick
		Vec4i l = NULL;
		l = detect_chopstic(foregroundMask, 70, 15, 100);
		// show live and wait for a key with timeout long enough to show images
		if (l[0] > 0 && l[1] > 0 && l[2] > 0 && l[3] > 0)
		{
			line(frame, Point(l[0], l[1]), Point(l[2], l[3]), Scalar(0, 255, 0), 2, 10);
			//draw dot at upper point
			int x = l[1] > l[3] ? l[2] : l[0];
			int y = l[1] > l[3] ? l[3] : l[1];
			circle(frame, Point(x, y), 5, Scalar(255, 0, 0), -1);
		}
		//detect spoon

		Vec3f c = NULL;
		c = detect_spoon(foregroundMask, 10, 30);
		if (c[0] > 0 && c[1] > 0 && c[2] > 0)
		{
			circle(frame, Point(c[0], c[1]), c[2], Scalar(0, 255, 0), 2);
			circle(frame, Point(c[0], c[1]), 2, Scalar(255, 0, 0), -1);
		}

		imshow("Live", frame);
		imshow("mask", foregroundMask);
		if (waitKey(5) >= 0)
			break;
	}
	return 0;
}
